[InstCombine] Enable saturated add canonicalization in more variants

AZero13 opened the pull request:
LLVM is not evaluating X u > C, a, b the same way it evaluates X <= C, b, a.

To fix this, let's move the folds to after the canonicalization of -1 to TrueVal.

Alive2 Proof:
https://alive2.llvm.org/ce/z/8QbZfx

AZero13 committed:
Pre-commit tests (NFC)

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,23 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
+  if (Pred == ICmpInst::ICMP_UGE &&
+      match(FVal, m_Add(m_Value(X), m_APInt(C))) && X == Cmp0 &&
+      match(TVal, m_AllOnes()) && match(Cmp1, m_APInt(CmpC)) && *CmpC == ~*C) {
+    // (X u >= ~C) ? -1 : (X + C) --> uadd.sat(X, C)
```
```suggestion
    // (X u>= ~C) ? -1 : (X + C) --> uadd.sat(X, C)
```

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,23 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
     return nullptr;
 
+  if (Pred == ICmpInst::ICMP_UGE &&
+      match(FVal, m_Add(m_Value(X), m_APInt(C))) && X == Cmp0 &&
```
```suggestion
      match(FVal, m_Add(m_Specific(Cmp0), m_APInt(C))) &&
```

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,23 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
 
+  if (Pred == ICmpInst::ICMP_UGE &&
+      match(FVal, m_Add(m_Value(X), m_APInt(C))) && X == Cmp0 &&
+      match(TVal, m_AllOnes()) && match(Cmp1, m_APInt(CmpC)) && *CmpC == ~*C) {
```
```suggestion
      match(Cmp1, m_APInt(CmpC)) && *CmpC == ~*C) {
```
Already checked above.

nikic reviewed (commented):
Comments apply to both copies of the code.

Can you please add an alive2 proof to the patch description?

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,32 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
+  if (Pred == ICmpInst::ICMP_UGE &&
+      match(FVal, m_Add(m_Specific(Cmp0), m_APInt(C))) &&
+      match(TVal, m_AllOnes()) && match(Cmp1, m_APInt(CmpC)) &&
+      *CmpC == (~*C + 1)) {
```
`~*C + 1` -> `-*C`.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,32 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
+  if (Pred == ICmpInst::ICMP_UGT &&
+      match(FVal, m_Add(m_Specific(Cmp0), m_APInt(C))) &&
+      match(TVal, m_AllOnes()) && match(Cmp1, m_APInt(CmpC)) &&
+      *CmpC == (~*C - 1)) {
```
nit: Here and above I would change the explicit `CmpC == F(C)` to use `m_SpecificInt`.

nikic reviewed (commented):
This PR has gone from handling 2 cases to 4 cases, but it doesn't seem like there is test coverage for all of them? (Please pre-commit the tests so it's visible that *all* the new code paths are actually needed.)

AtariDreams commented:
> This PR has gone from handling 2 cases to 4 cases, but it doesn't seem like there is test coverage for all of them? (Please pre-commit the tests so it's visible that *all* the new code paths are actually needed.)

Added cases for all of them!

AtariDreams commented:
> This PR has gone from handling 2 cases to 4 cases, but it doesn't seem like there is test coverage for all of them? (Please pre-commit the tests so it's visible that _all_ the new code paths are actually needed.)

Added cases for all of them! @nikic

AtariDreams commented:
@nikic @goldsteinn Is this adequate? I added 4 new test cases and 6 different tests were changed as a result of this.

goldsteinn commented:
Missing negative tests but code looks fine.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,30 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
+    // (X u> -C) ? -1 : (X + C) --> uadd.sat(X, C)
+    return Builder.CreateBinaryIntrinsic(Intrinsic::uadd_sat, Cmp0,
+                                         ConstantInt::get(Cmp0->getType(), *C));
+  }
```
Missing test coverage for this case. (If I comment it out, there are no failures.)

nikic replied:
Note that to hit UGE cases you need a test case where the select operands get swapped, otherwise you'll always get the canonical UGT form.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,30 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
+
+  if (Pred == ICmpInst::ICMP_UGE &&
+      match(FVal, m_Add(m_Specific(Cmp0), m_APInt(C))) &&
+      match(TVal, m_AllOnes()) && match(Cmp1, m_SpecificInt(-*C))) {
```
```suggestion
      match(Cmp1, m_SpecificInt(-*C))) {
```

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,30 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
+
+  if (Pred == ICmpInst::ICMP_UGT &&
+      match(FVal, m_Add(m_Specific(Cmp0), m_APInt(C))) &&
+      match(TVal, m_AllOnes()) && match(Cmp1, m_SpecificInt(~*C - 1))) {
```
```suggestion
      match(Cmp1, m_SpecificInt(~*C - 1))) {
```

AZero13 committed:
[InstCombine] Enable saturated add canonicalization in more variants

LLVM is not evaluating X u > C, a, b the same way it evaluates X <= C, b, a.

To fix this, let's move the folds to after the canonicalization of -1 to TrueVal.

Alive2 Proof:
https://alive2.llvm.org/ce/z/8QbZfx

AtariDreams commented:
@nikic Added more test cases for that and negative tests! @goldsteinn

AtariDreams commented:
@nikic I added missing test cases in addition to negative tests! @goldsteinn

AtariDreams commented:
@MaskRay Thoughts?

AtariDreams commented:
@nikic Is this good now?

AtariDreams commented:
@nikic Is this good now?

AtariDreams commented:
@goldsteinn I added missing test cases in addition to negative tests!

AtariDreams commented:
@spatel-gh Thoughts on this?

AtariDreams commented:
@nikic Ping.

AtariDreams commented:
@nikic Ping
